docs/getting-started/project-structure.md

```diff
@@ -64,3 +64,379 @@ config = rx.Config(
+and how often to refresh.
+
+```python
+import reflex as rx
+
+from ..models import Post
+from ..template import template
+from ..components.settings import SettingsState
+
+
+class PostsState(rx.State):
+    refresh_tick: int
+    page: int
+    posts: list[Post]
+
+    async def on_load(self):
+        settings = await self.get_state(SettingsState)
+        if settings.auto_update:
+            self.refresh_tick = settings.refresh_interval * 1000
+        else:
+            self.refresh_tick = 0
+
+    async def tick(self, _):
+        settings = await self.get_state(SettingsState)
+        with rx.session() as session:
+            q = Post.select().offset(self.page * settings.posts_per_page).limit(settings.posts_per_page)
+            self.posts = q.all()
+
+    def go_to_previous(self):
+        if self.page > 0:
+            self.page = self.page - 1
+
+    def go_to_next(self):
+        if self.posts:
+            self.page = self.page + 1
+
+
+@rx.page(route="/posts", on_load=PostsState.on_load)
+@template
+def posts():
+    return rx.vstack(
+        rx.foreach(PostsState.posts, post_view),
+        rx.hstack(
+            rx.button("< Prev", on_click=PostsState.go_to_previous),
+            rx.button("Next >", on_click=PostsState.go_to_next),
+            justify="between",
+        ),
+        rx.moment(interval=PostsState.refresh_tick, on_change=PostsState.tick, display="none"),
+        width="100%",
+    )
+```
+
+
+### Common State: `example_big_app/state.py`
+
+_Common_ states and substates that are shared by multiple pages or components
+should be implemented in a separate module to avoid circular imports. This
+module should not import other modules in the app.
+
+
+## Component Reusability
+
+The primary mechanism for reusing components in Reflex is to define a function that returns
+the component, then simply call it where that functionality is needed.
+
+Component functions typically should not take any State classes as arguments, but prefer
+to import the needed state and access the vars on the class directly.
+
+
+### Memoize Functions for Improved Performance
+
+In a large app, if a component has many subcomponents or is used in a large number of places, it can improve compile and runtime performance to memoize the function with the `@lru_cache` decorator.
+
+To memoize the `foo` component to avoid re-creating it many times simply add `@lru_cache` to the function definition, and the component will only be created once per unique set of arguments.
+
+```python
+from functools import lru_cache
+
+import reflex as rx
+
+class State(rx.State):
+    v: str = "foo"
+
+
+@lru_cache
+def foo():
+    return rx.text(State.v)
+
+
+def index():
+    return rx.flex(
+        rx.button("Change", on_click=State.set_v(rx.cond(State.v != "bar", "bar", "foo"))),
+        *[
+            foo()
+            for _ in range(100)
+        ],
+        direction="row",
+        wrap="wrap",
+    )
+```
+
+
+## External Components
```

Comment:
This should be a level 3 heading so that it nests under "Component Reusability"

Reply:
changed